R-package/tests/testthat/test_dataset.R

```diff
@@ -522,3 +525,24 @@ test_that("Dataset: method calls on a Dataset with a null handle should raise an
     dtrain$set_reference(reference = dvalid)
   }, regexp = "cannot get column names before dataset has been constructed")
 })
+
+test_that("lgb.Dataset$get_feature_num_bin() works", {
+  data <- matrix(nrow = 100L, ncol = 5L)
```

Comment:
> also the matrix creation is a bit cumbersome

I can help! You can use `data.matrix()` in R to extract the data from a `data.frame` into an R matrix object.

I'd do it like this:

```r
raw_df <- data.frame(
  all_random = runif(100L)
  , two_vals = rep(c(1L, 2L), 50L)
  , three_vals = c(rep(c(0L, 1L, 2L), 33L), 0L)
  , two_vals_plus_missing = c(rep(c(1L, 2L), 49L), NA_real_, NA_real_)
  , all_zero = rep(0L, 100L)
)
raw_mat <- data.matrix(raw_df)
```

Adding column names in the call to `data.frame()` isn't necessary, but I do think it's nice to have because it makes it a bit clearer what the intention of each column is.

> because in the end the matrix is numeric, not integer

An R matrix object is a multi-dimensional version of the R vector object. Both have the constraint that all data must be the same type.

So when you call `matrix()` or `as.matrix()` or `data.matrix()`, the resulting object will have all values of the same type, and R will do casting to get them all to the same common type. Given a mix of numeric and integer columns, the resulting matrix will be all numeric.

You can confirm that by checking the "storage mode" of the matrix.

```r
df <- data.frame(int_col = seq_len(10L), numeric_col = rnorm(10L))
sapply(df, class)
#    int_col numeric_col
#  "integer"   "numeric"

dmat <- data.matrix(df)
storage.mode(dmat)
# [1] "double"
```

In case a user presents an all-integer matrix, the LightGBM R package converts it to double before passing it to the C++ code used for Dataset construction:

https://github.com/microsoft/LightGBM/blob/1d8bfd8c9b3fa7d56db81201adedb61cfc11669c/R-package/R/lgb.Dataset.R#L80-L85

The difficulties in this process explain why #4323 would be really valuable to have.

Let me know if you have other questions! I appreciate you being willing to make these changes in the R package.

Reply:
awesome!

Could I add you to the R-specific parts of `CODEOWNERS`? In other words...are you open to helping with PR reviews on changes to the R package?

Reply:
Sure!

Reply:
Ok great, I'll put up a PR with that change. Thanks very much for that!

I think @StrikerRUS will be happy for the help, he has to be tired of reading my R code by now 😂